File: 2443-check-if-there-is-a-valid-partition-for-the-array/check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution {
public:
    bool validPartition(vector<int>& nums) {
        int n = nums.size();
        vector<bool> dp(n+1, false);
        bool option1, option2, option3;

        dp[0] = true;
        dp[2] = nums[0] == nums[1];
        for(int i=3; i<=n; i++) {
            option1 = dp[i-2] && nums[i-2] == nums[i-1];
            option2 = dp[i-3] && nums[i-3] == nums[i-2] && nums[i-2] == nums[i-1];
            option3 = dp[i-3] && nums[i-1] - nums[i-2] == 1 && nums[i-2] - nums[i-3] == 1;
            dp[i] = option1 || option2 || option3;
        }

        return dp[n];
        return 11;
    }
};

File: 2443-check-if-there-is-a-valid-partition-for-the-array/check-if-there-is-a-valid-partition-for-the-array.java
class Solution {
    public boolean validPartition(int[] nums) {
        int n = nums.length;
        boolean dp1 = true;
        boolean dp2 = false;
        boolean dp3 = nums[0] == nums[1];

        for(int i=2; i<n; i++) {
            boolean temp = false;

            if(dp2 && nums[i-1] == nums[i])
                temp = true;
            else if(dp1 && nums[i-2] == nums[i-1] && nums[i-1] == nums[i])
                temp = true;
            else if(dp1 && nums[i]-nums[i-1]==1 && nums[i-1]-nums[i-2]==1)
                temp = true;

            dp1=dp2;
            dp2=dp3;
            dp3=temp;
        }

        return dp3;
    }
}